HDFS-15843. Make write cross-platform

GauthamBanasandra opened the pull request:
* Currently write from unistd.h is used
  which isn't cross-platform.
* We need to use std::cout.write instead.

GauthamBanasandra committed:
HDFS-15843. Make write cross-platform

* Currently write from unistd.h is used
  which isn't cross-platform.
* We need to use std::cout.write instead.

GauthamBanasandra committed:
Use printf instead of write

* Replacing the plaform specific
  write method call with printf.

GauthamBanasandra commented:
@smengcl could you please review this PR?

jojochuang commented on hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfspp/examples/c/connect_cancel/connect_cancel.c:
```diff
@@ -43,10 +42,10 @@ const char *catch_exit   = "Exiting the signal handler.\n";
-  ssize_t res = write(1 /*posix stdout fd*/, msg, strlen(msg));
-  (void)res;
+  }
+  printf("%s", msg);
```
was this line added to debug and not needed?

GauthamBanasandra replied:
@jojochuang This line is needed. I've replaced `write` call that was writing to **stdout** with `printf`, so that it's cross platform.

GauthamBanasandra replied:
I've removed this. Please take a look at the latest commit.

smengcl commented:
Thanks @GauthamBanasandra for the patch.

I'm wondering if the [author](https://issues.apache.org/jira/browse/HDFS-11028) uses syscall write on purpose. According to the comment, heap could be corrupted if the `printf` call uses `malloc` or equivalent (depending on C/C++ library's implementation):

https://github.com/apache/hadoop/blob/8783461e2ec3aff2630ea3574a69beb1d5c61e84/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfspp/examples/c/connect_cancel/connect_cancel.c#L43

https://github.com/apache/hadoop/blob/869317be0a1fdff23be5fc500dcd9ae4ecd7bc29/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfspp/examples/cc/connect_cancel/connect_cancel.cc#L47-L48

How about wrapping the `<unistd.h>` header import line and the existing approach in `#ifndef WINDOWS`? And use `#else` to wrap the new approach. Assuming Windows is the platform you are aiming for.

GauthamBanasandra committed:
Add XPlatform WriteToStdout

* Added a method WriteToStdout
  under XPlatform::Syscall to
  handle cross-platform writes
  to the standard output stream.

GauthamBanasandra committed:
Fix warnings

* Removed blank tabs.

GauthamBanasandra commented:
Thanks for your review and suggestion @smengcl , @jojochuang. I've re-implemented my solution using the `write` system call for Windows and Linux separately, neatly tucked inside the `XPlatform` class 😁. Please do have another look at my PR.

GauthamBanasandra commented:
@smengcl could you please review my PR?

smengcl reviewed (approved):
Thanks @GauthamBanasandra !
